### ISSUE TITLE
Pravega Controller got restarted on running 10 streams with 10 KB data size 

### ISSUE THREAD

**Issue by SomeshJoshi19** [pravega/pravega#3898]
**Problem description**
On creating 10 parallel streams with IO workload supplied by each stream is: - 5 workers ( readers & writers ), 1 event/sec rate, 10 KB data size, run time 300 secs and 1000 segments, observed Pravega controller restart. 

The controller log before restart shows repeated occurrences of  `ReinitializationRequiredException`, `IllegalStateException` and `ConnectionLossException`.

```
2019-06-11 08:52:21,335 8429498 [Delegate] WARN  i.p.c.e.impl.EventProcessorCell - Restarting event processor: EventProcessor[scaleGroup:0] due to exception: {}
io.pravega.client.stream.ReinitializationRequiredException: io.pravega.client.stream.ReaderNotInReaderGroupException: Reader: 78656d0d-79d0-45a7-8b00-bb98b8fb4098
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:97)
        at io.pravega.controller.eventProcessor.impl.EventProcessorCell$Delegate.run(EventProcessorCell.java:103)
        at com.google.common.util.concurrent.AbstractExecutionThreadService$1$2.run(AbstractExecutionThreadService.java:66)
        at com.google.common.util.concurrent.Callables$4.run(Callables.java:119)
        at java.lang.Thread.run(Thread.java:748)
Caused by: io.pravega.client.stream.ReaderNotInReaderGroupException: Reader: 78656d0d-79d0-45a7-8b00-bb98b8fb4098
        at io.pravega.client.stream.impl.ReaderGroupStateManager.getCheckpoint(ReaderGroupStateManager.java:387)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.updateGroupStateIfNeeded(EventStreamReaderImpl.java:178)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:109)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:94)
        ... 4 common frames omitted
```

```
2019-06-11 08:52:23,523 8431686 [Delegate] INFO  i.p.c.s.i.AsyncSegmentInputStreamImpl - Closing reader for _system/_RGabortStreamReaders/0.#epoch.0
2019-06-11 08:52:23,408 8431571 [Delegate] WARN  i.p.c.e.impl.EventProcessorCell - Restarting event processor: EventProcessor[scaleGroup:0] due to exception: {}
java.lang.IllegalStateException: Reader is closed
        at com.google.common.base.Preconditions.checkState(Preconditions.java:507)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:92)
        at io.pravega.controller.eventProcessor.impl.EventProcessorCell$Delegate.run(EventProcessorCell.java:103)
        at com.google.common.util.concurrent.AbstractExecutionThreadService$1$2.run(AbstractExecutionThreadService.java:66)
        at com.google.common.util.concurrent.Callables$4.run(Callables.java:119)
        at java.lang.Thread.run(Thread.java:748)
2019-06-11 08:52:23,522 8431685 [Curator-Framework-0] ERROR o.a.c.f.imps.CuratorFrameworkImpl - Background operation retry gave up
org.apache.zookeeper.KeeperException$ConnectionLossException: KeeperErrorCode = ConnectionLoss
        at org.apache.zookeeper.KeeperException.create(KeeperException.java:99)
        at org.apache.curator.framework.imps.CuratorFrameworkImpl.checkBackgroundRetry(CuratorFrameworkImpl.java:844)
        at org.apache.curator.framework.imps.CuratorFrameworkImpl.performBackgroundOperation(CuratorFrameworkImpl.java:972)
        at org.apache.curator.framework.imps.CuratorFrameworkImpl.backgroundOperationsLoop(CuratorFrameworkImpl.java:925)
        at org.apache.curator.framework.imps.CuratorFrameworkImpl.access$300(CuratorFrameworkImpl.java:73)
        at org.apache.curator.framework.imps.CuratorFrameworkImpl$4.call(CuratorFrameworkImpl.java:322)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

Pravega-Operator deployment with 2 controllers, 3 Zookeepers, 4 Bookies and 3 Segment stores.
Environment details: PKS / K8 with medium cluster:
```
1 master nodes @ large.cpu (4 CPU, 4 GB Ram, 16 GB Disk)
3 worker nodes @ xlarge.cpu(4 cpu, 16 GB Ram, 32 GB Disk)
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
Pravega details:
```
Pravega version : 0.5.0-2269.6f8a820
Zookeeper Operator : pravega/zookeeper-operator:0.2.1
Pravega Operator: pravega/pravega-operator:0.3.2
```

**Comment by SomeshJoshi19** [comment]
logs for the issue
[new_log_20190611.zip](https://github.com/pravega/pravega/files/3276305/new_log_20190611.zip)

**Comment by sumit-bm** [comment]
Based on Triage discussion moving this issue to version 0.6.0

**Comment by RaulGracia** [comment]
In the Controller logs provided, I see frequent connection loss messages related to Zookeeper (even with large gaps without log info):
```
2019-06-11 08:59:14,047 8841639 [grpc-default-executor-334] INFO  i.p.c.s.r.g.v.ControllerServiceImpl - getURI called for segment NewScope/pps12d6/902.
2019-06-11 09:43:51,440 11519032 [ControllerServiceMain-SendThread(pravega-example-zk-0.pravega-example-zk-headless.default.svc.cluster.local:2181)] WARN  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 6803ms for sessionid 0x200ce37d6ea0020
2019-06-11 09:43:51,440 11519032 [ControllerServiceMain-SendThread(pravega-example-zk-0.pravega-example-zk-headless.default.svc.cluster.local:2181)] INFO  org.apache.zookeeper.ClientCnxn - Client session timed out, have not heard from server in 6803ms for sessionid 0x200ce37d6ea0020, closing socket connection and attempting reconnect
2019-06-11 09:43:51,542 11519134 [ControllerServiceMain-EventThread] WARN  o.a.c.retry.ExponentialBackoffRetry - Sleep extension too large (5000). Pinning to 1000
2019-06-11 09:43:51,546 11519138 [ControllerServiceMain-EventThread] INFO  org.apache.curator.ConnectionState - Connection string changed to: pravega-example-zk-0.pravega-example-zk-headless.default.svc.cluster.local:2181,pravega-example-zk-1.pravega-example-zk-headless.default.svc.cluster.local:2181,pravega-example-zk-2.pravega-example-zk-headless.default.svc.cluster.local:2181
2019-06-11 09:43:51,551 11519143 [ControllerServiceMain-EventThread] INFO  o.a.c.f.s.ConnectionStateManager - State change: SUSPENDED
```

This issue has been reported before the fix to the Zookeeper membership in here: https://github.com/pravega/zookeeper-operator/pull/69. The proof that the deployment of this issue suffered from the original Zookeeper membership problem is this (`Error writing next dynamic config file to disk`):
```
2019-06-11 06:14:53,911 [myid:1] - INFO  [QuorumPeer[myid=1](plain=/0.0.0.0:2181)(secure=disabled):Leader@432] - LEADING - LEADER ELECTION TOOK - 80 MS
2019-06-11 06:14:53,914 [myid:1] - INFO  [QuorumPeer[myid=1](plain=/0.0.0.0:2181)(secure=disabled):FileTxnSnapLog@367] - Snapshotting: 0x0 to /data/version-2/snapshot.0
2019-06-11 06:14:53,921 [myid:1] - ERROR [QuorumPeer[myid=1](plain=/0.0.0.0:2181)(secure=disabled):QuorumPeer@1564] - Error writing next dynamic config file to disk: 
2019-06-11 06:14:53,923 [myid:1] - INFO  [QuorumPeer[myid=1](plain=/0.0.0.0:2181)(secure=disabled):Leader@1264] - Have quorum of supporters, sids: [ [1],[1] ]; starting up and setting last processed zxid: 0x100000000
2019-06-11 06:14:53,924 [myid:1] - ERROR [QuorumPeer[myid=1](plain=/0.0.0.0:2181)(secure=disabled):QuorumPeer@1596] - Error closing file: 
2019-06-11 06:14:53,931 [myid:1] - INFO  [QuorumPeer[myid=1](plain=/0.0.0.0:2181)(secure=disabled):CommitProcessor@255] - Configuring CommitProcessor with 1 worker threads.
```

The previous fix (Zookeeper image available at `pravega/zookeeper:0.2.2` or `pravega/zookeeper:latest`) is critical for the correct behavior of the Zookeeper cluster. Many instability problems in the past were caused to this issue.

Moreover, in this report I don't see the resources provided to the Controller. If the experiment creates 1000 segments, it would be interesting to understand if the amount of provisioned CPU and RAM resources to Controller pods is enough or not.

Given that in recent times we have executed +7 days longevity experiments with much higher loads, I'm inclined to close this issue and re-open it again if it can be reproduced with i) the Zookeeper image fix and, ii) the right resource provisioning to Controllers.
